### PR TITLE
Add feature to increase the number of host to device transfer threads

### PR DESCRIPTION
**Pull request by chandrasekhard2**
This config improves the ResNet training performance on TPU v4 chips. 

```
Epoch 3 train begin 19:02:08
| Training Device=xla:0/1 Epoch=3 Step=5100 Loss=3.82812 Rate=2014.43 GlobalRate=1220.24 Time=19:02:14
| Training Device=xla:0/3 Epoch=3 Step=5100 Loss=3.39062 Rate=2014.43 GlobalRate=1223.24 Time=19:02:14
| Training Device=xla:0/0 Epoch=3 Step=5100 Loss=3.64062 Rate=2014.44 GlobalRate=1221.29 Time=19:02:14
| Training Device=xla:0/2 Epoch=3 Step=5100 Loss=3.34375 Rate=2014.61 GlobalRate=1221.54 Time=19:02:14
| Training Device=xla:0/0 Epoch=3 Step=5400 Loss=3.51562 Rate=2221.05 GlobalRate=1254.91 Time=19:02:30
| Training Device=xla:0/3 Epoch=3 Step=5400 Loss=3.40625 Rate=2221.03 GlobalRate=1256.85 Time=19:02:30
| Training Device=xla:0/1 Epoch=3 Step=5400 Loss=3.54688 Rate=2221.02 GlobalRate=1253.85 Time=19:02:30
| Training Device=xla:0/2 Epoch=3 Step=5400 Loss=3.43750 Rate=2221.59 GlobalRate=1255.17 Time=19:02:30
| Training Device=xla:0/1 Epoch=3 Step=5700 Loss=2.89062 Rate=2267.96 GlobalRate=1284.59 Time=19:02:47
| Training Device=xla:0/3 Epoch=3 Step=5700 Loss=3.50000 Rate=2267.92 GlobalRate=1287.56 Time=19:02:47
| Training Device=xla:0/0 Epoch=3 Step=5700 Loss=3.42188 Rate=2267.86 GlobalRate=1285.63 Time=19:02:47
| Training Device=xla:0/2 Epoch=3 Step=5700 Loss=3.45312 Rate=2268.25 GlobalRate=1285.89 Time=19:02:47
| Training Device=xla:0/0 Epoch=3 Step=6000 Loss=3.03125 Rate=2323.80 GlobalRate=1315.59 Time=19:03:03
| Training Device=xla:0/3 Epoch=3 Step=6000 Loss=3.32812 Rate=2323.76 GlobalRate=1317.51 Time=19:03:03
| Training Device=xla:0/1 Epoch=3 Step=6000 Loss=3.28125 Rate=2323.71 GlobalRate=1314.55 Time=19:03:03
| Training Device=xla:0/2 Epoch=3 Step=6000 Loss=3.37500 Rate=2323.66 GlobalRate=1315.84 Time=19:03:03
| Training Device=xla:0/1 Epoch=3 Step=6300 Loss=3.51562 Rate=2377.17 GlobalRate=1343.67 Time=19:03:19
| Training Device=xla:0/0 Epoch=3 Step=6300 Loss=3.12500 Rate=2377.16 GlobalRate=1344.70 Time=19:03:19
| Training Device=xla:0/3 Epoch=3 Step=6300 Loss=3.37500 Rate=2377.15 GlobalRate=1346.61 Time=19:03:19
| Training Device=xla:0/2 Epoch=3 Step=6300 Loss=3.12500 Rate=2377.08 GlobalRate=1344.95 Time=19:03:19
| Training Device=xla:0/3 Epoch=3 Step=6600 Loss=3.31250 Rate=2368.34 GlobalRate=1373.45 Time=19:03:35
| Training Device=xla:0/0 Epoch=3 Step=6600 Loss=3.20312 Rate=2368.35 GlobalRate=1371.56 Time=19:03:35
| Training Device=xla:0/1 Epoch=3 Step=6600 Loss=3.10938 Rate=2368.32 GlobalRate=1370.53 Time=19:03:35
| Training Device=xla:0/2 Epoch=3 Step=6600 Loss=3.04688 Rate=2367.79 GlobalRate=1371.79 Time=19:03:35
| Training Device=xla:0/3 Epoch=3 Step=6900 Loss=3.20312 Rate=2376.36 GlobalRate=1399.20 Time=19:03:51
| Training Device=xla:0/1 Epoch=3 Step=6900 Loss=3.20312 Rate=2376.36 GlobalRate=1396.30 Time=19:03:51
| Training Device=xla:0/0 Epoch=3 Step=6900 Loss=3.23438 Rate=2376.29 GlobalRate=1397.32 Time=19:03:51
| Training Device=xla:0/2 Epoch=3 Step=6900 Loss=3.26562 Rate=2376.85 GlobalRate=1397.57 Time=19:03:51
| Training Device=xla:0/2 Epoch=3 Step=7200 Loss=3.03125 Rate=2481.11 GlobalRate=1424.40 Time=19:04:06
| Training Device=xla:0/0 Epoch=3 Step=7200 Loss=3.18750 Rate=2480.62 GlobalRate=1424.14 Time=19:04:06
| Training Device=xla:0/3 Epoch=3 Step=7200 Loss=3.26562 Rate=2480.56 GlobalRate=1426.01 Time=19:04:06
| Training Device=xla:0/1 Epoch=3 Step=7200 Loss=3.17188 Rate=2480.55 GlobalRate=1423.12 Time=19:04:06
Epoch 3 train end 19:04:20
Epoch 3 test begin 19:04:20
| Test Device=xla:0/0 Step=0 Epoch=3 Time=19:04:21
| Test Device=xla:0/3 Step=0 Epoch=3 Time=19:04:22
| Test Device=xla:0/1 Step=0 Epoch=3 Time=19:04:22
| Test Device=xla:0/2 Step=0 Epoch=3 Time=19:04:22
Epoch 3 test end 19:04:28, Accuracy=31.63

### CODE REVIEW

File: test/test_train_mp_imagenet.py [hunk 1]
```diff
@@ -76,12 +91,39 @@
 
 DEFAULT_KWARGS = dict(
     batch_size=128,
-    test_set_batch_size=64,
+    test_set_batch_size=128,
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet.py:94]: does this work on v3 too?

Comment (chandrasekhard2) [reply test/test_train_mp_imagenet.py:94]: It works on v3 but I haven't checked on v2. I will revert this.

File: test/test_train_mp_imagenet.py [hunk 2]
```diff
@@ -81,7 +96,34 @@
...
+    num_workers=8,
+    host_to_device_transfer_threads=1,
 )
+'''
+Best config to achieve peak performance on v4-8
+
+  1. It is recommended to use this config in conjuntion with XLA_USE_BF16=1 Flag. 
+  2. Hyperparameters can be tuned to further improve the accuracy.
+'''
+OPTIMIZED_KWARGS = dict(
```
Comment (yeounoh) [inline test/test_train_mp_imagenet.py:112]: This is great, thanks @chandrasekhard2

**Review by JackCaoG — APPROVED** [review]
mostly lGTM, can we post a more detail fake data + real data benchmark with and without this pr? Does this impact convergence?

File: test/test_train_mp_imagenet.py (hunk 2 above)
Comment (miladm) [reply test/test_train_mp_imagenet.py:112]: NOTE as future reference: the configs that work best on v4-8 are not expected to work as well on other hardware configs (e.g. v4-128)

@chandrasekhard2 wdyt about re-naming your `OPTIMIZED_KWARGS` variable? See below:

`OPTIMIZED_KWARGS_v4_8 = dict(...`
`OPTIMIZED_KWARGS = OPTIMIZED_KWARGS_v4_8`

**Review by miladm — CHANGES_REQUESTED** [review]
Thanks @chandrasekhard2 

Added a couple of comments.

Also, can you please document (in this PR) how much performance gain should we expect to emerge from this PR?

**Comment by chandrasekhard2** [comment]
> Thanks @chandrasekhard2
> 
> Added a couple of comments.
> 
> Also, can you please document (in this PR) how much performance gain should we expect to emerge from this PR?

@miladm - This config improves the performance even on v4-128. Its just that there is still a gap if we compare it to Optimized tensorflow ResNet.

File: test/test_train_mp_imagenet.py (hunk 2 above)
Comment (will-cromar) [reply test/test_train_mp_imagenet.py:112]: I agree with renaming this to clarify the accelerator this was optimized for. If it was optimized for multiple v4 sizes, then we can name it `OPTIMIZED_KWARGS_v4`.

If we're going to start publishing optimized hyperparameters for specific accelerators, we should move them to some sort of config file system, e.g. [gin](https://github.com/google/gin-config).

The corresponding [Flax](https://github.com/google/flax/tree/main/examples/imagenet) and [TensorFlow](https://github.com/tensorflow/models/tree/master/official/vision) examples both implement something like this.

**Comment by JackCaoG** [comment]
@chandrasekhard2 Let's see we can address review comments and run some experiments before EOD tmr, we will cut next rc soon and we only have ~ 2 week left before release.

File: test/test_train_mp_imagenet.py [hunk 3]
```diff
@@ -94,8 +136,10 @@
                 'lr_scheduler_type': 'WarmupAndExponentialDecayScheduler',
             })
 }
-
-# Set any args that were not explicitly given by the user.
+'''
```
[the remark this answers is not in the document]
Comment (chandrasekhard2) [reply test/test_train_mp_imagenet.py]: Thanks for pointing. I fixed it.

File: test/test_train_mp_imagenet.py [hunk 4]
```diff
@@ -81,7 +96,33 @@
...
+    loader_prefetch_size=8,
+    device_prefetch_size=4,
+    num_workers=8,
+    host_to_device_transfer_threads=1,
 )
+
+#  Best config to achieve peak performance on TPU v4
+#    1. It is recommended to use this config in conjuntion with XLA_USE_BF16=1 Flag.
+#    2. Hyperparameters can be tuned to further improve the accuracy.
+
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet.py:110]: remove empty line maybe?

File: test/test_train_mp_imagenet.py [hunk 5]
```diff
@@ -81,7 +96,33 @@
...
+OPTIMIZED_KWARGS_v4 = dict(
+    batch_size=128,
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet.py:112]: I thought we can do 256 for v4?

File: test/test_train_mp_imagenet.py [hunk 6]
```diff
@@ -81,7 +96,33 @@
...
+    device_prefetch_size=4,
+    num_workers=8,
+    host_to_device_transfer_threads=1,
 )
+
+#  Best config to achieve peak performance on TPU v4
+#    1. It is recommended to use this config in conjuntion with XLA_USE_BF16=1 Flag.
+#    2. Hyperparameters can be tuned to further improve the accuracy.
+
+OPTIMIZED_KWARGS_v4 = dict(
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet.py:111]: how do user run this test with v4 kwargs? Can it be done by a flag?

File: test/test_train_mp_imagenet.py (hunk 5 above)
Comment (chandrasekhard2) [reply test/test_train_mp_imagenet.py:112]: Step time would jump from 39ms to 82ms if we increase the batch size from 128 to 256. (3ms more)

File: test/test_train_mp_imagenet.py (hunk 6 above)
Comment (chandrasekhard2) [reply test/test_train_mp_imagenet.py:116]: Added a flag and sample command to show how to run.

File: test/test_train_mp_imagenet.py [hunk 7]
```diff
@@ -38,7 +38,25 @@
...
+        'type': int,
+    },
+    '--device_prefetch_size': {
+        'type': int,
+    },
+    '--host_to_device_transfer_threads': {
+        'type': int,
+    },
+    '--use_optimized_kwargs': {
+        'action': 'store_true',
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet.py:58]: can we make a `str`? so user can use it like `--use_optimized_kwargs tpuv4` and verifiy if it works?

Comment (chandrasekhard2) [reply test/test_train_mp_imagenet.py:58]: Yeah I can do that.

**Review by JackCaoG — APPROVED** [review]

**Comment by JackCaoG** [comment]
CI fialure is unrelated, rebase would solve it.

**Review by miladm — APPROVED** [review]
Thanks @chandrasekhard2! LGTM.
